lib/core/src/Cardano/Wallet/Primitive/Scripts.hs

```diff
@@ -0,0 +1,118 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+-- |
+-- Copyright: © 2018-2020 IOHK
+-- License: Apache-2.0
+--
+-- An implementation of wallet's verification key discovery
+-- in a script.
+
+module Cardano.Wallet.Primitive.Scripts
+    ( isShared
+    , retrieveAllVerKeyHashes
+    ) where
+
+import Prelude
+
+import Cardano.Address.Script
+    ( KeyHash (..), Script (..), ScriptHash (..), toScriptHash )
+import Cardano.Crypto.Wallet
+    ( XPub )
+import Cardano.Wallet.Primitive.AddressDerivation
+    ( Depth (..)
+    , DerivationType (..)
+    , Index (..)
+    , Role (..)
+    , SoftDerivation (..)
+    )
+import Cardano.Wallet.Primitive.AddressDerivation.Shelley
+    ( ShelleyKey (..) )
+import Cardano.Wallet.Primitive.AddressDiscovery.Sequential
+    ( SeqState (..)
+    , VerificationKeyPool (..)
+    , defaultAddressPoolGap
+    , getAddressPoolGap
+    , mkUnboundedAddressPoolGap
+    , mkVerificationKeyPool
+    )
+import Cardano.Wallet.Primitive.Types.Address
+    ( AddressState (..) )
+import Control.Monad
+    ( foldM )
+import Data.Functor.Identity
+    ( Identity (..) )
+
+import qualified Data.List as L
+import qualified Data.Map.Strict as Map
+
+instance Ord ScriptHash where
+    compare (ScriptHash sh1) (ScriptHash sh2) = compare sh1 sh2
+
+isShared
+    :: (k ~ ShelleyKey, SoftDerivation k)
+    => Script
+    -> SeqState n k
+    -> ([k 'ScriptK XPub], SeqState n k)
+isShared script (SeqState !s1 !s2 !ixs !rpk !prefix !scripts !s3) =
+    let verKeysInScript = retrieveAllVerKeyHashes script
+        (VerificationKeyPool accXPub currentGap verKeyMap) = s3
+        projectKey (ShelleyKey k) = ShelleyKey k
+        toVerKey = deriveAddressPublicKey accXPub MultisigScript
+        projectIndex = toEnum . fromInteger . toInteger . getIndex @'Soft
+        ourVerKeysInScript =
+            map (\(_,(ix,_)) -> toVerKey (projectIndex ix) ) $
+            filter (\(keyH,_) -> keyH `elem` verKeysInScript)
+            (Map.toList verKeyMap)
+        scriptXPubs = L.nub $ map projectKey ourVerKeysInScript
+        updateAddressState k current =
```

Comment:
We should consider dropping the 'addressState' entirely when it comes to script addresses. It does not have the same meaning at all than for standard payment address and I don't think there's any business case for it.

Reply:
resolved in private communication. We will use AddressState in order to know what keys were used (ie. discovered). And later expose to user